### DISCUSSION TITLE
Modulo 2. No puedo ejecutar el comando para crear el entorno virtual en mi carpeta seleccionada.

### DISCUSSION THREAD

**Discussion by JossPerezT** [LaunchX-InnovaccionVirtual/CursoIntroPython#93] [Q&A]
Hola. Buenas tardes. 
Estoy ejecutando el módulo 2 para crear el entorno virtual, pero no puedo crearlo en los archivos de mis katas. 
![image](https://user-images.githubusercontent.com/84040594/153515079-74af0da6-1151-4fd0-9e29-ce454e7dc099.png)

Pero al ejecutarlo en **C:\Windows\System32**, me deja perfectamente hacerlo:
![image](https://user-images.githubusercontent.com/84040594/153514881-ed098123-2a10-4b93-901a-85a6e81f5d18.png)

¿Cómo puedo hacer que se ejecute en ciertas carpetas?

**Comment by GabrielVillagran** [comment]


**Reply by JossPerezT** [reply]
Gracias! 
Lo he intentado pero tampoco me funciona :(

**Comment by MSuarez0** [comment]
esto se debe a que el no encuentra el comando python, intenta con python3, o tambien en la instalacion hay que poner el PATCH. te dejo una refencia
![win_installer](https://user-images.githubusercontent.com/99224742/153520030-d04f0181-80cc-4076-b5d5-cf51aeb408b2.png)

**Reply by MarioBerne** [reply]
Para el path como puedo hacerlo? o lo más rápido sería desinstalar y volver a instalar Python?

**Comment by Dfredude** [comment]
Hola Joss
- Eso sucede porque Python no esta en tus **_Path variables_**
Corre `python --version` en tu cmd, estando en el directorio del **_Modulo 2_** y te debe salir algo asi:
![image](https://user-images.githubusercontent.com/72330887/153519990-0750173e-c0b5-47b3-9c15-814f203e403f.png)
De lo contrario, tienes que agregar Python al path, googlealo o aqui esta un video 
https://www.youtube.com/watch?v=0bDRUOpec4c

**Comment by doguedogue** [comment]
Si te das cuenta te funcionó con python pero no con python3.
puedes hacerlo en cualquier carpeta, en tu caso se está creando la carpeta del entorno virtual: env.

Prueba este ejercicio: el entorno virtual debe ser independiente de tu entorno global, por eso una vez creado el entorno virtual, entras y ves los pips instalados, se prueba instalar y/o desinstalar la biblioteca para encriptar, la prueba consiste en ver que funcione en uno y en otro no.

```
#******** Entorno virtual
#1. instalar  virtualenv
pip install virtualenv

#2.  crear entorno virtual Python (con esta si me funcionó en windows)
$ virtualenv env

#2.1 Python 3 (no me funcionó en windows)
#$ python3 -m venv env

#2.2 Python (no me funcionó en windows)
#$ python -m venv env

#3. Activar entorno virtual (windows)
env\Scripts\activate

#4. Ver PIPs
pip freeze

#5. Instalar/desinstalar pip de prueba
#pip uninstall bcrypt
pip -q install bcrypt

#6. probar comando adentro de entorno virtual
python -c "import bcrypt; print(bcrypt.hashpw('password'.encode('utf-8'), bcrypt.gensalt()))"

#7. Salir del entorno virtual
deactivate

#8. probar comando afuera de entorno virtual
python -c "import bcrypt; print(bcrypt.hashpw('password'.encode('utf-8'), bcrypt.gensalt()))"
```

**Comment by JossPerezT** [comment]
Gracias a todos. 
Intenté modificar path y los ejercicios. 
Desafortunadamente, dentro de la carpeta de mis katas sigo sin poder instalar el entorno virtual. 
Sin embargo en otras carpetas si puedo, así que haré mi ejercicio por ahí. 

Gracias a todos por su tiempo y consejos. 😁

**Comment by 3rickDJ** [comment]
# Los espacios de las carpetas :performing_arts: 
Hey Joss! Aquí Erick, podrías intentar lo siguiente y decirme qué te muestra la consola :smile_cat: 

Probablemente lo que esté pasando aquí (no he investigado a profundidad) es que **python** puede llegar a tener problemas con direcciones de archivos que contengan espacios. En tu caso si te das cuenta te funciona el comando en la carpeta `C:\WINDOWS\system32`, notarás que en esta dirección no existe ningún espacio, como sí lo hay en `.\VS code\On Boarding\Modulo 2`. Poner espacios sería como poner un acento en `Modulo`, y ya sabrás que eso puede llevar a conflictos.


![image](https://user-images.githubusercontent.com/94492584/153529267-585dc15a-44fc-47d4-ac96-bbc9d220b18f.png)

### Te recomiendo cambies los nombres de tus carpetas y archivos de tal forma que no contengan espacios o caracteres especiales (raros jaja). Y probar de nuevo el comando

Inténtalo con: 
1. `python3 -m venv env`
2.  `python -m venv env`


**Déjame saber si lo solucionaste** :mage_man:

**Reply by JossPerezT** [reply]
Hola Erick. 
Creí que si iba a ser mi solución, pero no 😓
![image](https://user-images.githubusercontent.com/84040594/153537802-d38903fa-2a58-4909-a92b-9287ca345640.png)


Creo que debe ser un problema dentro de Desktop, porque hasta el usuario me permite utilizar el entorno virtual.

**Comment by ArielCanedo** [comment]
Prueba así 
En la barra del directorio de la carpeta donde estás, escribe CMD con la ruta de la carpeta y prueba
python -m venv env
ve si se crea la carpeta, si sí
Ya puedes usar el comando 
env\Scripts\activate

Podría funcionar

**Reply by JossPerezT** [reply]
A menos que te haya entendido mal, no me funcionó. 
![image](https://user-images.githubusercontent.com/84040594/153546709-77822651-af0c-4610-8d17-c761d6ae33d5.png)

**Comment by BarajasAngel** [comment]
### Posible solución

Si no cuentas con otras versiones de Python instaladas puede que se deba a un problema de Windows defender ya que suele hacer que no se puedan crear carpetas desde cmd desde ciertos directorios para hacer que la carpeta donde quieres crear el entorno virtual no sea afectada por eso puedes seguir los siguientes pasos. Una vez hayas terminado los pasos también recomiendo abrir cmd como administrador solo le das botón derecho abrir como administrador, solo para asegurar que no falten permisos a la hora de ejecutar el comando.

### 1. Busca Windows Defender o Seguridad de Windows en el buscador

![image](https://user-images.githubusercontent.com/99161300/153550862-6801bcfd-15d5-4a8a-ad75-c447324a9ce1.png)

### 2. Una vez abierto dale click en la parte protección contra virus y amenazas

![1](https://user-images.githubusercontent.com/99161300/153550981-4868ad5e-3c91-4ec7-9a0f-55f87d0e2a58.png)
 de 

### 3. Luego te iras a configuración de protección contra virus y amenazas

![2](https://user-images.githubusercontent.com/99161300/153551061-4f347100-5f39-412c-9100-c0de85fc5b36.png)

### 4, Una vez ahí bajaras hasta encontrar la opción de exclusiones

![3](https://user-images.githubusercontent.com/99161300/153551173-89413bd5-2297-4a1f-bd13-dcbf1172f01b.png)

### 5. Aquí deberás agregar la ruta de la carpeta donde quieres crear el entorno virtual

![4](https://user-images.githubusercontent.com/99161300/153551360-a3e975de-5ad3-42dc-8fb0-053d8ce35013.png)

Aparentemente el error que estas teniendo es por una confusión en el path por tener mas de una versión de python sin embargo en el caso de que solo tengas una versión esto podría ser el causante debido a que Windows 10 agrego la protección de carpetas para evitar que malware no intente crear carpetas con arcvhivos malicioso y comprometer tu computadora, de funcionar podría explicar porque te deja crear el entorno en system32 y no donde querías. Espero te funcione.

**Reply by JossPerezT** [reply]
Hola, gracias por el tutorial. 
Tampoco me funcionó y solo tengo una versión de python. 
![image](https://user-images.githubusercontent.com/84040594/153671955-cfffdc73-a830-47c9-a618-d0655b532012.png)
Y sigue marcando el mismo error 😅 
![image](https://user-images.githubusercontent.com/84040594/153672059-b984a027-78dc-4763-b534-b2161c5d8849.png)

**Comment by LeonardoGutierrezV** [comment]
Buenos dias.

Intenta con lo siguiente.
Manten presionada la tecla **windows** y presiona la tecla **r** deberia mostrarse una ventana similar a esta:
![image](https://user-images.githubusercontent.com/88008394/153586331-27630693-5aad-4b6a-b588-48e16b4972da.png)
teclea cmd y valida en que carpeta se esta abriendo

![image](https://user-images.githubusercontent.com/88008394/153586451-00d45290-2879-4a17-b239-ca16be47a672.png)

Para mi caso es la carpeta del usuario.

A continuación intenta ejecutar tu python
solo escribe python y presiona **enter** debería lucir así

![image](https://user-images.githubusercontent.com/88008394/153586703-019f7779-1bce-45aa-9fc7-560b54b31942.png)

Si te marca un error eso nos indica que no tienes configuradas algunas variables del sistema como  indican arriba, si se ejecuta de manera normal trata de hacer lo siguiente.

teclea en python **exit()** y presiona enter

eso te regresara donde comenzaste

![image](https://user-images.githubusercontent.com/88008394/153587021-69db16a2-2a3e-4798-bdc3-6a63e4cdaa01.png)


En tu explorador de windows ubica la carpeta donde quieres guardar tu entorno, da clic en la barra de direcciones y copia la ubicación.

da clic en la barra de direcciones 
![image](https://user-images.githubusercontent.com/88008394/153587458-6b8edaf1-8beb-4d2d-80e7-8c59d95865a8.png)

y copia esa ubicación

ahora regresa a tu cmd  escribe **cd**  agrega un espacio y da clic en el icono de la parte superior derecha de esta ventana

![image](https://user-images.githubusercontent.com/88008394/153587748-8de8be7b-4bb8-4cd1-a8d9-a783e845e663.png)

ve a la opcion editar y despues a la opcionpegar y deberia quedar asi 

![image](https://user-images.githubusercontent.com/88008394/153589159-b5f99b9a-a39b-4d9a-b99f-2bbab97091f0.png)

presiona enter y estaras en la carpeta que necesitas utilizar

![image](https://user-images.githubusercontent.com/88008394/153589254-6c34ccec-1900-4eb2-84ba-823151c89f3b.png)

ahora teclean: **python3 -m venv env**

en su defecto 

ahora teclean: **python -m venv env**

a mi me funciono el segundo


![image](https://user-images.githubusercontent.com/88008394/153589883-cf37060e-61e7-44c7-b77f-bc89f804bc82.png)

y se habra creado tu entorno virtual

![image](https://user-images.githubusercontent.com/88008394/153589933-deab3f81-6601-4ee2-a83c-8dc4ed209d6b.png)

Ahora lo que mencionan de las variables del entorno o del sistema operativo las verificas presionando la tecla **windows y R**

y tecleas **sysdm.cpl**
![image](https://user-images.githubusercontent.com/88008394/153590463-c44056c3-b266-44e0-abc5-f88ed8c90f4f.png)

presionas enter y te mostrara la siguiente ventana

![image](https://user-images.githubusercontent.com/88008394/153590580-4b7a7070-5ae6-4ce9-abf3-37a1db75ace1.png)
****
pica en opciones avanzadas y luego en el botón variables de  entorno

![image](https://user-images.githubusercontent.com/88008394/153590820-cd1e494c-7014-4c47-8634-915ce54434bf.png)


y deberia verse mas o menos asi

![image](https://user-images.githubusercontent.com/88008394/153590985-bd3d185b-55eb-4a04-ae6f-96cfad880c38.png)

 en el apartado path debería estar registrada la ubicación donde tienes instalado python.

Si picas el botón editar deberia poder agregar la dirección so no esta en la lista 

![image](https://user-images.githubusercontent.com/88008394/153591158-247b133e-6887-43b3-ae4d-f64b38ced930.png)

**Reply by IrmaDiazA** [reply]
Hola Leonardo, aunque yo no fui la que preguntó, te agradezco mucho tu explicación detallada ya que yo también tenía muchos problemas con esta kata y me ayudaste mucho. Mil gracias! :)

**Reply by Meii21** [reply]
Muchas gracias Leo :)

**Reply by LeonardoGutierrezV** [reply]
Para eso estamos!!!
Saludos y éxito!!!

**Reply by JossPerezT** [reply]
Hola Leo. Gracias por el tutorial. 
Según lo que entiendo mis variables están bien, en la consola me reconoce python e hice lo que me indicaste e igual me marca ese "Error". 
![image](https://user-images.githubusercontent.com/84040594/153673348-72a48ae3-3dd7-41d0-a1e7-6806e5b87e6e.png)

Mis variables de entorno igual están configuradas. 
![image](https://user-images.githubusercontent.com/84040594/153673471-3b4dd78b-0d11-4389-b312-f6d055ffe70f.png)
La primera contiene la carpeta general de python y la segunda la de scripts. A menos que haya hecho algo mal. 
El entorno virtual lo termine creando en **C:\Users\LIZZETT** que es en donde si me deja.

**Reply by RIvanCF** [reply]
Me ayudo muchísimo tu tutorial, muchas gracias

**Comment by Juan-Cruz15** [comment]
![screenshot_20220211_161719](https://user-images.githubusercontent.com/99099757/153678193-f1e1890d-c047-45d2-9603-f6825ca225cc.png)

Alguien sabrá por que me aparece así

**Reply by LeonardoGutierrezV** [reply]
No esta instalado Python, O es tema de tus variables de entorno!

**Comment by DanielGalindoRios** [comment]
intenta solo escribiendo py  en vez de python o python3... a mi me funciono.

![image](https://user-images.githubusercontent.com/99298545/153734248-47489185-f2ca-4e94-8164-9c5657b22dfd.png)

**Reply by JossPerezT** [reply]
Gracias. 
No, py no lo reconoce 😟
